driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java

```diff
@@ -107,30 +113,76 @@ public final class CommandMessage extends RequestMessage {
+        List<ByteBuf> byteBuffers = bsonOutput.getByteBuffers();
+        try {
+            CompositeByteBuf byteBuf = new CompositeByteBuf(byteBuffers);
+            try {
+                byteBuf.position(getEncodingMetadata().getFirstDocumentPosition());
+                ByteBufBsonDocument byteBufBsonDocument = createOne(byteBuf);
+
+                // If true, it means there is at least one Kind 1:Document Sequence in the OP_MSG
+                if (byteBuf.hasRemaining()) {
+                    BsonDocument commandBsonDocument = byteBufBsonDocument.toBaseBsonDocument();
+
+                    // Each loop iteration processes one Document Sequence
+                    // When there are no more bytes remaining, there are no more Document Sequences
+                    while (byteBuf.hasRemaining()) {
+                        // skip reading the payload type, we know it is 1
+                        byteBuf.position(byteBuf.position() + 1);
+                        int sequenceStart = byteBuf.position();
+                        int sequenceSizeInBytes = byteBuf.getInt();
+                        int sectionEnd = sequenceStart + sequenceSizeInBytes;
+                        
+                        String fieldName = getSequenceIdentifier(byteBuf);
+                        // If this assertion fires, it means that the driver has started using document sequences for nested fields.  If
+                        // so, this method will need to change in order to append the value to the correct nested document.
+                        assertFalse(fieldName.contains("."));
+
+                        ByteBuf documentsByteBufSlice = byteBuf.duplicate().limit(sectionEnd);
+                        try {
+                            commandBsonDocument.append(fieldName, new BsonArray(createList(documentsByteBufSlice)));
+                        } finally {
+                            documentsByteBufSlice.release();
+                        }
+                        byteBuf.position(sectionEnd);
+                    }
+                    return commandBsonDocument;
+                } else {
+                    return byteBufBsonDocument;
+                }
+            } finally {
+                byteBuf.release();
+            }
+        } finally {
+            byteBuffers.forEach(ByteBuf::release);
         }
-
-       return commandBsonDocument;
     }
 
-    boolean containsPayload() {
-        return payload != null;
+    /**
+     * Get the field name from a buffer positioned at the start of the document sequence identifier of an OP_MSG Section of type
+     * Document Sequence (Kind 1).
+     * <p>
+     * Upon normal completion of the method, the buffer will be positioned at the start of the first BSON object in the sequence.
+    */
+    private String getSequenceIdentifier(final ByteBuf byteBuf) {
+        ByteArrayOutputStream sequenceIdentifierBytes = new ByteArrayOutputStream();
```

Comment:
thanks for adopting my suggestion! A further improvement is to give a good guess of 'default size' to avoid resizing again and again, but it is out of my knowledge scope to do that. If no clue, it is good enough.

Reply:
The default for ByteArrayOutputStream is 32, which is large enough for all known sequence identifiers.